src/colorizer/AnnotationData.ts

```diff
@@ -0,0 +1,222 @@
+    for (let i = 0; i < this.labelData.length; i++) {
+      if (this.labelData[i].ids.has(id)) {
+        labelIdxs.push(i);
+      }
+    }
+    return labelIdxs;
+  }
+
+  getLabeledIds(labelIdx: number): number[] {
+    this.validateIndex(labelIdx);
+    return Array.from(this.labelData[labelIdx].ids);
+  }
+
+  getTimeToLabelIdMap(dataset: Dataset): Map<number, Record<number, number[]>> {
+    if (this.timeToLabelIdMap !== null) {
+      return this.timeToLabelIdMap;
+    }
+
+    const timeToLabelIdMap = new Map<number, Record<number, number[]>>();
+
+    for (let labelIdx = 0; labelIdx < this.labelData.length; labelIdx++) {
+      const ids = this.labelData[labelIdx].ids;
+      for (const id of ids) {
+        const time = dataset.times?.[id];
+        if (time === undefined) {
+          continue;
+        }
+        if (!timeToLabelIdMap.has(time)) {
+          timeToLabelIdMap.set(time, {});
+        }
+        if (!timeToLabelIdMap.get(time)![labelIdx]) {
+          timeToLabelIdMap.get(time)![labelIdx] = [];
+        }
+        timeToLabelIdMap.get(time)![labelIdx].push(id);
+      }
+    }
+    this.timeToLabelIdMap = timeToLabelIdMap;
+    return timeToLabelIdMap;
+  }
+
+  // Setters
+
+  /** Creates a new label and returns its index. */
+  createNewLabel(name?: string, color?: Color): number {
+    if (!color) {
+      const palette = KNOWN_CATEGORICAL_PALETTES.get(DEFAULT_CATEGORICAL_PALETTE_KEY)!;
+      color = new Color(palette.colorStops[this.numLabelsCreated % palette.colorStops.length]);
+    }
+    if (!name) {
+      name = `Label ${this.numLabelsCreated + 1}`;
+    }
+
+    this.labelData.push({
+      name: name,
+      color: color,
+      ids: new Set(),
+    });
+
+    this.numLabelsCreated++;
+    return this.labelData.length - 1;
+  }
+
+  private validateIndex(idx: number): void {
+    if (idx < 0 || idx >= this.labelData.length) {
+      throw new Error(`Invalid label index: ${idx}`);
+    }
+  }
+
+  setLabelName(labelIdx: number, name: string): void {
+    this.validateIndex(labelIdx);
+    this.labelData[labelIdx].name = name;
+  }
+
+  setLabelColor(labelIdx: number, color: Color): void {
+    this.validateIndex(labelIdx);
+    this.labelData[labelIdx].color = color;
+  }
+
+  deleteLabel(labelIdx: number): void {
+    this.validateIndex(labelIdx);
+    this.labelData.splice(labelIdx, 1);
+    this.timeToLabelIdMap = null;
+  }
+
+  applyLabelToId(labelIdx: number, id: number): void {
+    this.validateIndex(labelIdx);
+    this.labelData[labelIdx].ids.add(id);
+    this.timeToLabelIdMap = null;
+  }
+
+  removeLabelFromId(labelIdx: number, id: number): void {
+    this.validateIndex(labelIdx);
+    this.labelData[labelIdx].ids.delete(id);
+    this.timeToLabelIdMap = null;
+  }
+
+  toggleLabelOnId(labelIdx: number, id: number): void {
+    if (!this.labelData[labelIdx].ids.has(id)) {
+      this.applyLabelToId(labelIdx, id);
+    } else {
+      this.removeLabelFromId(labelIdx, id);
+    }
+  }
```

Comment:
This is also optional, but I'm just a tiny bit hesitant about these methods, particularly `toggleLabelOnId`, whose behavior is dependent on state which the caller doesn't actually observe. It's less "set this data" and more "take this action," if that makes sense - it maps well onto what the UI does, but not necessarily onto this class's role as a data container. I might consider something like `setLabelOnId(labelIdx: number, id: number, applied: boolean)` and `getIdHasLabel(labelIdx: number, id: number): boolean`. Toggling could then be performed like `setLabelOnId(labelIdx, id, !getIdHasLabel(labelIdx, id))`. (As a bonus, this means fewer setter methods to pass through `wrapFunctionInUpdate`.)